### PR TITLE
Themes Discovery: Add track events

### PR DESCRIPTION
**Pull request by rcrdortiz**
Related to:
- https://github.com/Automattic/dotcom-forge/issues/4307

We need to add tracking capabilities to the Discovery carousels' cards and navigation component.

Note: After the PR gets merged we will need to add the events to Tracks following the PCYsg-mUj-p2, the issue hasn't been created yet since the new event names might change.

## Proposed Changes

* Refactored the tracking architecture outside of the ThemesSelection component.
* Refactored the ThemesSelection component to use the new Tracking mechanism.
* Unified the tracking event properties so that all events include the same properties.
* Reduced the complexity of the ThemesSelection component thanks to the new tracking mechanism.
* Added tracking capabilities to the Collection component cards and navigation.

## Testing Instructions

* Open the Calypso Live Link.
* Open Production on another tab.
* Click on a theme in the showcase (Non-collections)
* Compare the event across the proposal and production.
* The properties on the proposal should include all the production properties. In some cases, there will be new properties added so that all events contain the same properties. The filterString is replaced by three props, tier, filter and search.
* Now open the LoTS using the feature flag `?flags=themes/discovery-lots`
* When clicking on a card and on the style variation buttons events should be registered.
* When clicking on the `See all` button or navigation arrows, events should be registered.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/themes/collections/showcase-theme-collection.tsx [hunk 1]
```diff
@@ -26,8 +29,36 @@ export default function ShowcaseThemeCollection( {
...
+
+	const onStyleVariationClick = (
+		themeId: string,
+		resultsRank: number,
+		variation: { slug: string }
+	) => {
+		recordThemeClick( themeId, resultsRank, 'style_variation', variation?.slug );
+		variation
+			? recordThemeStyleVariationClick( themeId, resultsRank, '', variation.slug )
+			: recordThemesStyleVariationMoreClick( themeId, resultsRank );
```
Comment (rcrdortiz) [inline client/my-sites/themes/collections/showcase-theme-collection.tsx:90]: These are the same events we're firing on the Theme Showcase. I think that it doesn't make much sense if at all to have them. For starters, we fire a `calypso_themeshowcase_theme_click` with whatever is going on. And then we fire another event with basically the same information but a different name.

I think we could simplify this by just changing the action of the main event to either `style_variation` or `more_style_variations` and we would have the same result but we could reduce the amount of code needed for this.

I checked if the events are being used in funnels and the `calypso_themeshowcase_theme_style_variation_more_click` isn't used at all and the `calypso_themeshowcase_theme_style_variation_click` is only used in one funnel. So I think it's safe to delete these events, but I just wanted a sanity check from you all. If we delete them, the `ThemeSelection` and `ThemeCollection` components would be less complex.

If no one has anything against this I'm going to remove these "duplicate" events either on this PR or in a follow-up.

File: client/my-sites/themes/events/theme-showcase-tracks.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+			collectionId: string | null = null,
+			collectionIndex: number | null = null
+		) => {
+			const adjustedPosition = themePosition + 1;
+			const isCollection = null !== defaultCollectionId || null !== collectionIndex;
+
+			recordTracksEvent( event, {
+				search: query?.search,
+				tier: query?.tier,
+				filter: query?.filter,
```
Comment (rcrdortiz) [inline client/my-sites/themes/events/theme-showcase-tracks.ts:47]: Before we were recording the `filterString` which isn't always available and depends on the client building it based on the `query` prop. So I just changed this to show the actual query instead which has the same information (actually more since the tier wasn't included in the filter string).

File: client/components/theme-collection/index.tsx [hunk 3]
```diff
@@ -101,13 +120,19 @@ export default function ThemeCollection( {
 					<div className="theme-collection__description">{ preventWidows( description ) }</div>
 				</div>
 				<div className="theme-collection__carousel-controls">
-					<Button className="theme-collection__see-all" onClick={ onSeeAll }>
+					<Button className="theme-collection__see-all" onClick={ onSeeAllAction }>
 						{ translate( 'See all' ) }
 					</Button>
-					<Button className="theme-collection__carousel-nav-button theme-collection__carousel-nav-button--previous">
+					<Button
+						onClick={ () => trackNavigationClick( 'left' ) }
```
Comment (dsas) [inline client/components/theme-collection/index.tsx]: I think it's better to use prev/next rather than left/right. It's just slightly less ambiguous when considering RTL languages.

File: client/components/theme-collection/index.tsx [hunk 4]
```diff
@@ -101,13 +120,19 @@ export default function ThemeCollection( {
 					<div className="theme-collection__description">{ preventWidows( description ) }</div>
 				</div>
 				<div className="theme-collection__carousel-controls">
-					<Button className="theme-collection__see-all" onClick={ onSeeAll }>
+					<Button className="theme-collection__see-all" onClick={ onSeeAllAction }>
 						{ translate( 'See all' ) }
 					</Button>
-					<Button className="theme-collection__carousel-nav-button theme-collection__carousel-nav-button--previous">
+					<Button
```
Comment (dsas) [inline client/components/theme-collection/index.tsx:126]: Should there be a specific event for when you hit the end and rewind the carousel? It kind of shows that people were interested in viewing more.

Maybe a nice to have.

Comment (rcrdortiz) [reply client/components/theme-collection/index.tsx:126]: Not sure about the business value of it. We have the see all button which does have an event.

File: client/components/theme-collection/index.tsx (hunk 3 above)
Comment (rcrdortiz) [reply client/components/theme-collection/index.tsx]: Sure. Done

**Review by dsas — COMMENTED** [review]
Weirdly when clicking on the meatball menu for my active theme on production I get `is_active_theme: false`, on this pr I get `is_active_theme: true` 👍 

I also see that `search_term` and `search_taxonomies` have been replaced `search`, `tier` and `filter` 👍 

❗  One difference I noticed with the discovery lots is that `tier` was set to `marketplace` when I clicked on a marketplace theme in the partner collection, even though the search etc was not set. This is different to the lots without discovery where tier is left blank when clicking on a marketplace theme. Tier is also blank when using discovery lots and clicking on a marketplace theme for other carousels like portfolio.

File: client/my-sites/themes/events/theme-showcase-tracks.ts (hunk 2 above)
Comment (Copons) [reply client/my-sites/themes/events/theme-showcase-tracks.ts:47]: Maybe I'm misunderstanding this but, since we are extracting an existing functionality, let's not change its behaviour in the same PR.

I'm fine adding new props for the collections, but not about changing existing ones.

If it's necessary to change them to make it work with collections, then I'd rather split this task in 2:
1. Extract the tracking and apply it to existing use cases.
2. Apply it to the new (collection) use cases.

Besides, what if some funnels (or other data analyses) rely on the current props?

It's generally safer to add a new property (and describe the difference clearly in the event details) — but then again, I'm not exactly sure what is changing here — which is kind of my point. 😅 

I've taken a screen of the `calypso_themeshowcase_theme_click` prop on prod vs dev (clicking on the Fewer theme from the `/filter/portfolio` URL):

| Prod | Dev |
|--------|--------|
| Screenshot 2023-10-26 at 12 06 37 | Screenshot 2023-10-26 at 12 06 53 | 

We are removing several props (e.g. `search_taxonomies`, `search_term`), and replacing them with others (e.g. `filter`, `search`, `tier`).

What about keeping both for now?

File: client/my-sites/themes/collections/showcase-theme-collection.tsx (hunk 1 above)
Comment (Copons) [reply client/my-sites/themes/collections/showcase-theme-collection.tsx:90]: According to Tracks, both these events are owned by Lego. I'd check with them first. 🙂 

(I'd also definitely address this in a follow-up, as I've already rambled at length in another comment: https://github.com/Automattic/wp-calypso/pull/83456/files/7804c4b3885bf1f663d360f07309f86e0c983d81#r1372986824 😄 )

File: client/components/theme-collection/index.tsx (hunk 4 above)
Comment (Copons) [reply client/components/theme-collection/index.tsx:126]: I think it's enough to track the nav buttons themselves and "See all" to determine engagement. 🤔 
And I agree with Richard that clicking on "See all" should (ideally at least) directly answer the question "are people interested in viewing more?".

File: client/my-sites/themes/events/theme-showcase-tracks.ts (hunk 2 above)
Comment (rcrdortiz) [reply client/my-sites/themes/events/theme-showcase-tracks.ts:47]: > What about keeping both for now?

I would need to drill down the function that calculates `search_taxonomies` and `search_terms`. Both of those fields are kinda of the same and don't include the `tier` of searches. I think that the current implementation is a bit lacking since it's more complex than my proposal and includes less information than my proposal. 

But I can restore the search taxonomies and rank for the non-collection events and in a follow-up remove it + the duplicated events.

So to recap, I'll restore those props and send them empty on the collection's view. I'll leave the new props in this PR, then in a follow-up speak with Lego about removing the search props they've created and the duplicated events.

File: client/my-sites/themes/collections/showcase-theme-collection.tsx (hunk 1 above)
Comment (rcrdortiz) [reply client/my-sites/themes/collections/showcase-theme-collection.tsx:90]: Let's continue the conversation in that comment then. Resolving this comment.

**Comment by rcrdortiz** [comment]
> Weirdly when clicking on the meatball menu for my active theme on production I get `is_active_theme: false`, on this pr I get `is_active_theme: true` 👍
> 
> I also see that `search_term` and `search_taxonomies` have been replaced `search`, `tier` and `filter` 👍
> 
> ❗ One difference I noticed with the discovery lots is that `tier` was set to `marketplace` when I clicked on a marketplace theme in the partner collection, even though the search etc was not set. This is different to the lots without discovery where tier is left blank when clicking on a marketplace theme. Tier is also blank when using discovery lots and clicking on a marketplace theme for other carousels like portfolio.

I see your concern, in this implementation proposal tier is not actually the tier of the theme, but the tier of the search. That is if you changed the `view` filter on the LoTS. I'll have a look and see if I can get the actual tier of the theme (which might be really challenging)

**Comment by dsas** [comment]
> I see your concern, in this implementation proposal tier is not actually the tier of the theme, but the tier of the search. That is if you changed the view filter on the LoTS. 

No, weirdly that is what was happening. I had no search or tier set and this proposal was setting `tier` to marketplace but only when I had collections enabled and used the partner collection.

**Comment by dsas** [comment]
Here's a video in case that's clearer:

https://github.com/Automattic/wp-calypso/assets/93301/f6e27f66-8dfc-47ff-b737-f3d17cc20276

**Comment by rcrdortiz** [comment]
> > I see your concern, in this implementation proposal tier is not actually the tier of the theme, but the tier of the search. That is if you changed the view filter on the LoTS.
> 
> No, weirdly that is what was happening. I had no search or tier set and this proposal was setting `tier` to marketplace but only when I had collections enabled and used the partner collection.

Thanks, I'll have a look and see what's going on with the tier.

File: client/my-sites/themes/events/theme-showcase-tracks.ts (hunk 2 above)
Comment (Copons) [reply client/my-sites/themes/events/theme-showcase-tracks.ts:47]: Thanks! 🙇‍♂️ 

> I think that the current implementation is a bit lacking since it's more complex than my proposal and includes less information than my proposal.

I agree!
It's just to make our changes more predictable, and hopefully avoid regressions on code-adjacent tasks, like data analysis.

**Comment by rcrdortiz** [comment]
> Thanks, I'll have a look and see what's going on with the tier.

@dsas I had a look and everything is working as intended. The event information is filled by the theme container and/or filters, and not with the theme data (just like what we currently have in prod). For clicks on the tier dropdown we have thisi event `calypso_themeshowcase_filter_pricing_click`. The tier prop in the `calypso_themeshowcase_theme_click` event serves as a way to identify where the click originated from and it's position in the list. Some examples:
- Discover LoTS, click on a theme in a collection => the data is filled in with the collection information and isCollection = true
- Discover LoTS, click on a theme in the collection view => this case has the exact same data as the previous but with isCollection = false since it isn't triggered from a collection carousel.

File: client/my-sites/themes/events/theme-showcase-tracks.ts (hunk 2 above)
Comment (rcrdortiz) [reply client/my-sites/themes/events/theme-showcase-tracks.ts:47]: Restored the `search_taxonomies` and `search_term` event props. These props are now also added to all the triggered events (in prod they aren't available for all three events).

**Comment by dsas** [comment]
> The tier prop in the calypso_themeshowcase_theme_click event serves as a way to identify where the click originated from and it's position in the list.

Isn't that what `is_collection`, `collection` and `collection_index` are for?

I understand that the tier prop is not to indicate the type of theme (`theme_type` already exists and you said above it's not filled with theme data). 

### LOTS discovery behaviour

Tier is empty when I am using the Featured carousel, but populated when I use the Partner carousel. 

When I use 'See all' from a marketplace carousel and then click a theme, `tier` is set and `is_collection` is false with the other collection properties missing).

When I use 'See all' from the recommended carousel and then click a partner theme, `tier` is  empty.

----

It seems like tier is pointless from a carousel list, it's sometimes blank and the collection related properties are more useful.

I'm not sure that it's useful from a 'see all' list. It tells us that someone clicked a 'marketplace' theme on a 'partner' see all list but not if someone clicked a 'marketplace' theme on a differet kind of 'see all' list.


I think there's probably something I'm missing?

**Comment by dsas** [comment]
The restored search_term and search_taxonomies are working as they do in production...

...though production seem buggy in that they both remember previous searches - if I search for `subject:blog` manually then `subject:restaurant` and then `coffee` I get:

key | value
----|-------
search_term | subject:blog subject:restaurant subject:restaurant coffee
search_taxonomies | subject:blog subject:restaurant subject:restaurant

But as I say, this happens on live too so lets not sweat it here.

**Comment by rcrdortiz** [comment]
> Isn't that what `is_collection`, `collection` and `collection_index` are for?
> 
> I understand that the tier prop is not to indicate the type of theme (`theme_type` already exists and you said above it's not filled with theme data).
> 
> ### LOTS discovery behaviour
> Tier is empty when I am using the Featured carousel, but populated when I use the Partner carousel.
> 
> When I use 'See all' from a marketplace carousel and then click a theme, `tier` is set and `is_collection` is false with the other collection properties missing).
> 
> When I use 'See all' from the recommended carousel and then click a partner theme, `tier` is empty.
> 
> It seems like tier is pointless from a carousel list, it's sometimes blank and the collection related properties are more useful.
> 
> I'm not sure that it's useful from a 'see all' list. It tells us that someone clicked a 'marketplace' theme on a 'partner' see all list but not if someone clicked a 'marketplace' theme on a differet kind of 'see all' list.
> 
> I think there's probably something I'm missing?

I also had some of those issues with how it works. What helped me realize the intent of these things is that the events don't originate from the theme, they're triggered from the containers of the themes. **So the sole purpose is to identify the container that was clicked** and what theme/position etc the theme has within that container, that means that the information used for building the events is obtained mainly from the theme showcase props and not from the actual theme (except for a few properties). And from what you're saying you're seeing it as if the theme was the primary actor here (I click on a marketplace theme but the tier doesn't get populated).

You also need to take into consideration that there exist some views that will never have the is_collection prop, for example, if you click on the `Business` (non-collection, you can imagine that collections don't even exist if you prefer) filter button you will see all the Business themes and an event will trigger with the base information if you click on a theme. Without these changes, if you're on the Business screen and decide to filter by `Partner` themes, since there is no tier prop, it is impossible to know if the click originated from a container that has business + any filter (current production behavior). With the tier changes we can now determine if the tier filter has a value and aid us in nailing down the container the click originated from.

**Review by dsas — APPROVED** [review]
Ok yeah, it does work as you say and it does logically make sense. 

It would be good if this was clearer, I guess the tracks event registration docs need to be updated anyway.

**Comment by dsas** [comment]
Thanks for your patience and explanation!

**Comment by rcrdortiz** [comment]
> Thanks for your patience and explanation!

No problem, thank you for challenging it and making me articulate exactly how it works. It should be much easier to document now and thanks to the conversation we're all on the same page so we all win 🙇
